Add elasticsearch message advisor bot

gale2307 opened the pull request.

aphedges reviewed (changes requested):
Please rebase this PR against `main` to resolve conflicts and make it easier to review. You should be able to see in the Git history how I made changes to `requirements.txt` and `__init__.py`, but I can answer any questions you have about how to fix things.

I'll look at this PR properly once you've done that.

aphedges commented:
Please also make sure that all checks pass both on your local machine and in CI.

gale2307 commented:
I'm very sorry it took me so long to reply - this pr completely slipped off my mind over the past few weeks. I'll rebase and make the necessary changes, will get back to you hopefully by sometime this week.

gale2307 commented:
Hi Alex, I rebased and made the necessary changes to make the bot compatible with the changes in main. Checks are also all green now. 

Additionally, I simplified the elasticsearch implementation in baseline_models to be able to restore data from snapshots instead of having to repopulate the index using the dataset. Hopefully this helps in testing the code.

aphedges commented on src/chiron_utils/bots/__init__.py:
```diff
@@ -31,6 +34,7 @@
         [
             LrAdvisor,
             LrPlayer,
+            ElasticAdvisor,
```
This list should be sorted:

```suggestion
            ElasticAdvisor,
            LrAdvisor,
            LrPlayer,
```

aphedges commented on src/chiron_utils/bots/elastic_bot.py:
```diff
@@ -0,0 +1,78 @@
+
+    is_first_messaging_round = False
+    player_type = diplomacy_strings.NO_PRESS_BOT
+    elastic_client = AutoencoderClient(ELASTIC_HOST, MODEL_PATH)
```
This accidentally shares the client across all instances of the bot. Instantiate the client on bot instantiation instead:

```suggestion
	from dataclasses import field
    elastic_client = field(default_factory=lambda: AutoencoderClient(ELASTIC_HOST, MODEL_PATH))
```

To understand the purpose of this change, see [Mutable default values — dataclasses — Data Classes — Python 3.11.11 documentation](https://docs.python.org/3.11/library/dataclasses.html#mutable-default-values).

aphedges reviewed (changes requested):
Please try to rebase this again. You accidentally pulled in some of my changes, which will make the changes you've made more difficult to see.

As far as I'm aware, your bot can't act as a player because it doesn't generate moves. If that is correct, then please merge `ElasticAdvisor` into `ElasticBot`, which should be renamed to `ElasticAdvisor`.

Remove the extra empty lines added in the README, if possible. I would guess that Prettier is adding them. If that is the case, try removing the empty line before your new `ElasticBot` section.

Once you've made these changes, I'll try the bot out!

gale2307 commented:
Thank you for the feedback! I addressed the issues you mentioned, let me know if there are anything else that needs to be changed.

gale2307 commented:
Got it, what issues did you have setting it up? I'll update the instructions before making a PR on the baseline-models repo.

aphedges commented:
> Got it, what issues did you have setting it up? I'll update the instructions before making a PR on the baseline-models repo.

Two main things:

- It's not clear what files from the snapshot zip file should be put in `src/baseline_models/message_advisor_code/.snapshots`.
- The code blocks should not have a leading `$ ` to make copying and pasting commands easier.

In my personal opinion, this would be easier to discuss directly on the `baseline-models` code, hence my request for you to make the PR first.

aphedges commented on src/chiron_utils/bots/elastic_advisor.py:
```diff
@@ -0,0 +1,74 @@
+# `localhost` doesn't work when running an agent with Docker Desktop
+if os.environ.get("VIRT") == "docker":
+    ELASTIC_HOST = re.sub(r"\blocalhost\b", "host.docker.internal", ELASTIC_HOST)
+ELASTIC_INDEX = "tagged_documents_masked_scaled_center"
```
Where does this value come from? If it's from `baseline-models`, it should be declared there instead.

aphedges commented on src/chiron_utils/bots/elastic_advisor.py:
```diff
@@ -0,0 +1,74 @@
+        for other_power in get_other_powers([self.power_name], self.game):
+            if other_power in messages:
+                if self.bot_type == BotType.ADVISOR:
+                    for i in range(min(MESSAGE_ADVICE_COUNT, len(messages[other_power]))):
+                        await self.suggest_message(other_power, messages[other_power][i])
```
You should be able to simplify this:

```suggestion
                    for message in messages[other_power][:MESSAGE_ADVICE_COUNT]:
                        await self.suggest_message(other_power, message)
```

For an example in action:

```python
>>> list(range(5))[:10]
[0, 1, 2, 3, 4]
>>> list(range(10))[:10]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
>>> list(range(20))[:10]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
```

aphedges commented on src/chiron_utils/bots/elastic_advisor.py:
```diff
@@ -0,0 +1,74 @@
+                if self.bot_type == BotType.ADVISOR:
+                    for i in range(min(MESSAGE_ADVICE_COUNT, len(messages[other_power]))):
+                        await self.suggest_message(other_power, messages[other_power][i])
+                elif self.bot_type == BotType.PLAYER:
+                    await self.send_message(other_power, messages[other_power][0])
```
You hardcode `bot_type = BotType.ADVISOR` earlier, so this conditional isn't needed. You can assume that the bot is always an advisor.

aphedges reviewed (changes requested):
Sorry for taking a week to get back to this, but I've looked over this PR and <https://github.com/ALLAN-DIP/baseline-models/pull/10>, and I've gotten everything to work! Once you make the remaining changes, we can hopefully merge them soon.

gale2307 commented:
I think this error occurs when querying elasticsearch while its still restoring data from the snapshot. I've updated the restore snapshot script in https://github.com/ALLAN-DIP/baseline-models/pull/10 to happen in the foreground to make this interaction clearer.

aphedges reviewed (approved):
This looks all good to me! Thanks for working with me to resolve everything and get it merged!

Once CI passes, I'll go ahead and merge.
